src/ophyd_async/core/_providers.py

```diff
@@ -1,12 +1,25 @@
 from abc import abstractmethod
 from dataclasses import dataclass
-from typing import Protocol, Sequence
+from pathlib import Path
+from typing import Optional, Protocol, Sequence, Union
 
 
 @dataclass
 class DirectoryInfo:
-    directory_path: str
-    filename_prefix: str
+    """
+    Information about where and how to write a file.
```

Comment:
If you can come up with a sufficiently brief explanation of why both `root` and `resource_dir` are needed, I think it would be useful to put in this docstring. Maybe something like

> The bluesky event model splits the URI for a resource into two segments to aid in different applications mounting filesystems at different mount points.